### DISCUSSION TITLE
VRP with specialised capacity constraints

### DISCUSSION THREAD

**Discussion by matthewrj** [google/or-tools#2964] [Routing (and legacy CP) questions]
I am trying to use or-tools to solve a complicated vehicle routing problem.

The problem is VRP with time windows, capacity and pickups and deliveries. The complicated part is the capacity. They have 14 compartments for the product organised into 4 decks, a lower deck and an upper deck in the truck, each with 3 compartments and a lower deck and upper deck in the trailer each with 4 compartments.

Each compartment has an area capacity eg. 2 m2 (they aren't uniform)
The truck has a max payload eg. 12 tonnes
The trailer has a max payload eg. 18 tonnes
The truck and trailer combined has a max payload eg. 28 tonnes
The payload of the upper deck of the truck has to be less than or equal to the payload of the lower deck
Same for the trailer
The gross mass of the trailer has to be less than or equal to 1.5 * the gross mass of the truck

Each product has a weight and area per unit eg. 300 kgs and 0.5 m2
You can sometimes mix products for different pickups and deliveries from different species in one compartment eg. one compartment can contain 1 unit of product A and 2 units of product B that were picked up from different locations and are being delivered to different locations, but you cannot put product A with product C in the same compartment.
You can only load/unload from a compartment if there is a clear path to it from the back of the truck ie. LIFO per deck. eg. you cannot load into or unload from compartment 01 if compartments 02 or 03 have products destined for a different location.

I have tried to model the quantity in each compartment using ConstraintSolver::MakePathCumul where nexts are taken from the routing model and creating my own cumuls and transits. I then do the same for the area capacity in each compartment and the payload of each deck. I then link the parallel transits to be multiples of the quantity transit vector
ie. for each (compartmentQtyTransitVar, compartmentAreaTransitVar) pair
```
solver.addConstraint(
  solver.makeEquality(solver.makeProd(compartmentQtyTransitVar, job.product.areaM2),
                      compartmentAreaTransitVar))
```

I then use MakeEquality to set the sum of all the compartment qty transit vars at a particular node to equal the number of units being picked up or delivered (negative if delivered).
```
solver.addConstraint(
  solver.makeEquality(solver.makeSum(allQtyTransitVars), job.quantity))
```

My issue is when I try run the solver, it does not cause the pickups and deliveries to be reordered  in such a way that these constraints are satisfied, it just fails to find a solution if the shortest path is to visit a series of pickups that cause the vehicle to exceed its capacity. 
eg. I solve for two pickups and deliveries and the shortest path is p1 -> p2 -> d2 -> d1, if p1 + p2 fit on the truck and satisfy the capacity constraints then all is well however if p1 + p2 exceed the constraints, instead of outputting p1 -> d1 -> p2 -> d2 it just fails to find a solution.

I am guessing there is all sorts of magic in RoutingDimension that is missing from my crude equivalent. I was under the impression you were able to add arbitrary constraints to the model by adding constraints to the ConstraintSolver instead of doing everything through the RoutingModel. My questions are 1. is there a better way to model this by doing everything through the RoutingModel and not dropping down into the ConstraintSolver? or if not, 2. what am I missing to make this output the solution p1 -> d1 -> p2 -> d2?

**Comment by matthewrj** [comment]
I've added my manually created transit variables to the finalizer via `routing.addVariableMaximizedByFinalizer`. Reading this line: https://github.com/google/or-tools/blob/39f44709bba203f5ff3bc18fab8098739f189a6d/ortools/constraint_solver/routing.h#L2922
Does this mean these variables will not cause the filter to fail and so not cause backtracking?

**Comment by matthewrj** [comment]
It works if I set FirstSolutionStrategy to FIRST_UNBOUND_MIN_VALUE. However, it is quite slow.

**Comment by matthewrj** [comment]
Using FIRST_UNBOUND_MIN_VALUE as a FirstSolutionStrategy produces very poor, unusable results. Disabling the specialised capacity constraints and using other strategies produces high quality results but are unusable because they can be infeasible to execute in the real world because of the specialised capacity constraints. Is there any way to use PARALLEL_CHEAPEST_INSERTION with arbitrary constraints built using the constraint solver API?

**Comment by geocept** [comment]
Looks as if this issue is open for more than 2 years now? I would be very interessted in a solution for such specialised capacity constraints, they are very common for all kinds transport jobs. Would be very nice if you could think about a solution in the near future.
